File: tests/python/test_tract_smoke.py
"""Smoke tests for the local tract (pure-Rust ONNX) embedding backend.

These tests verify that the typed PyO3 binding surface for the tract
provider is importable and that the data-carrier types behave correctly
without requiring a real model download.

The tract backend is behind a Cargo feature gate. If the wheel was not
built with ``--features tract``, the entire module is skipped.

Tract is a drop-in equivalent of :class:`FastEmbedModel` for environments
where the prebuilt ONNX Runtime binaries that back fastembed cannot be
linked (musl-libc Linux distributions, sandboxed targets, ...). It loads
the same fastembed model catalog via tract-onnx instead.

Tests that require a real model download are gated on
``BLAZEN_TEST_TRACT=1`` (and optionally ``BLAZEN_TRACT_MODEL_NAME`` to
override the default ``BGESmallENV15`` variant) to avoid surprise
multi-MB downloads in CI.

Run manually:

    BLAZEN_TEST_TRACT=1 \\
        uv run --no-sync pytest tests/python/test_tract_smoke.py -v
"""

import os

import pytest

# The full tract typed surface lives behind the `tract` Cargo feature.
# If the installed wheel was not built with that feature, skip the entire
# module rather than failing with an ImportError.
try:
    from blazen import (
        TractEmbedModel,
        TractError,
        TractOptions,
        TractResponse,
    )
except ImportError:
    pytest.skip(
        "blazen was not built with the tract feature "
        "(rebuild with: maturin develop -m crates/blazen-py/Cargo.toml --features tract)",
        allow_module_level=True,
    )

TRACT_ENABLED = os.environ.get("BLAZEN_TEST_TRACT") == "1"
MODEL_NAME = os.environ.get("BLAZEN_TRACT_MODEL_NAME")

skip_without_flag = pytest.mark.skipif(
    not TRACT_ENABLED,
    reason="BLAZEN_TEST_TRACT=1 not set",
)


# ---------------------------------------------------------------------------
# Pure typed-surface tests (no model required, always run)
# ---------------------------------------------------------------------------


def test_tract_options_typed():
    """TractOptions accepts the documented keyword-only fields and round-trips."""
    opts = TractOptions(
        model_name="BGESmallENV15",
        cache_dir="/tmp/blazen-tract-cache",
        max_batch_size=32,
        show_download_progress=True,
    )
    assert opts.model_name == "BGESmallENV15"
    assert opts.cache_dir == "/tmp/blazen-tract-cache"
    assert opts.max_batch_size == 32
    assert opts.show_download_progress is True


def test_tract_options_defaults_are_none():
    """TractOptions with no kwargs has all-None fields."""
    opts = TractOptions()
    assert opts.model_name is None
    assert opts.cache_dir is None
    assert opts.max_batch_size is None
    assert opts.show_download_progress is None


def test_tract_options_setters_mutate_inplace():
    """TractOptions exposes setters for each field."""
    opts = TractOptions()
    opts.model_name = "BGESmallENV15"
    opts.max_batch_size = 64
    opts.show_download_progress = False
    opts.cache_dir = "/var/cache/blazen"
    assert opts.model_name == "BGESmallENV15"
    assert opts.max_batch_size == 64
    assert opts.show_download_progress is False
    assert opts.cache_dir == "/var/cache/blazen"


def test_tract_embed_model_class_shape():
    """TractEmbedModel is a class with the expected method/getter surface."""
    # Constructing the model eagerly downloads/loads ONNX weights, so we
    # can't instantiate without a real model. Introspect the type instead.
    assert hasattr(TractEmbedModel, "embed")
    assert hasattr(TractEmbedModel, "model_id")
    assert hasattr(TractEmbedModel, "dimensions")


def test_tract_response_class_shape():
    """TractResponse exposes embeddings + model getters as a frozen class.

    Per crates/blazen-py/src/providers/tract.rs, ``TractResponse`` is a
    ``frozen`` pyclass with no Python-side ``__new__``: it is constructed
    only via ``TractEmbedModel.embed(...)``. Verify the attribute surface
    without instantiating.
    """
    assert hasattr(TractResponse, "embeddings")
    assert hasattr(TractResponse, "model")


def test_tract_error_is_exception():
    """TractError is an exception type subclassing the provider error hierarchy."""
    assert isinstance(TractError, type)
    assert issubclass(TractError, Exception)


# ---------------------------------------------------------------------------
# Live tests gated behind BLAZEN_TEST_TRACT=1 (downloads model on first run)
# ---------------------------------------------------------------------------


@skip_without_flag
def test_tract_embed_model_constructs_with_real_model():
    """TractEmbedModel(options=opts) succeeds and exposes model_id/dimensions."""
    opts = TractOptions(model_name=MODEL_NAME) if MODEL_NAME else TractOptions()
    model = TractEmbedModel(options=opts)
    assert isinstance(model.model_id, str)
    assert len(model.model_id) > 0
    assert isinstance(model.dimensions, int)
    assert model.dimensions > 0


@skip_without_flag
@pytest.mark.asyncio
async def test_tract_embed_returns_typed_response():
    """A real tract model produces a TractResponse with embeddings + model id."""
    opts = TractOptions(model_name=MODEL_NAME) if MODEL_NAME else TractOptions()
    model = TractEmbedModel(options=opts)

    response = await model.embed(["hello", "world"])

    # TractResponse round-trips its data-carrier surface.
    assert isinstance(response, TractResponse)
    assert isinstance(response.model, str)
    assert len(response.model) > 0

    assert len(response.embeddings) == 2
    assert all(len(v) == model.dimensions for v in response.embeddings)
    # Different inputs should produce different vectors.
    assert response.embeddings[0] != response.embeddings[1]


@skip_without_flag
@pytest.mark.asyncio
async def test_tract_embed_vector_dimensionality():
    """Each embedding vector has exactly model.dimensions elements."""
    opts = TractOptions(model_name=MODEL_NAME) if MODEL_NAME else TractOptions()
    model = TractEmbedModel(options=opts)

    response = await model.embed(["a single sentence"])

    assert len(response.embeddings) == 1
    assert len(response.embeddings[0]) == model.dimensions

File: tests/node/test_tract_smoke.mjs
// Tract pure-Rust ONNX embedding typed-binding smoke tests.
//
// Exercises the typed surface only, without downloading or running a model.
// The classes are exposed by blazen-node whenever the binding is compiled
// with the `embed-tract` feature; when the feature is absent the symbols
// are missing from the module exports and every test below silently
// passes through (early return) instead of failing.
//
// Build with the feature enabled first:
//     cd crates/blazen-node && npm install && npm run build -- --features embed-tract
//
// Run shape-only tests:
//     node --test tests/node/test_tract_smoke.mjs
//
// Run live tests (downloads / loads an ONNX model from Hugging Face):
//     BLAZEN_TEST_TRACT=1 node --test tests/node/test_tract_smoke.mjs
//
// All test bodies are synchronous: async tests hang the Node test runner
// because of the napi-rs tokio-handle lifecycle (project memory).

import test from "ava";

import * as blazen from "../../crates/blazen-node/index.js";

const TRACT_ENABLED = process.env.BLAZEN_TEST_TRACT === "1";

// Whether the native binding was built with the `embed-tract` feature. We
// probe for `TractEmbedModel` since it is the canonical entry point for the
// feature; if it is absent every other typed class will also be absent.
const HAS_TRACT = typeof blazen.TractEmbedModel === "function";

test("Tract typed bindings · TractOptions is an object-literal interface (no runtime class)", (t) => {
  if (!HAS_TRACT) {
    t.pass("embed-tract feature not built");
    return;
  }
  // `JsTractOptions` is a TypeScript interface -- at runtime it is just a
  // plain object passed to `TractEmbedModel.create`. Confirm there is no
  // accidental class export shadowing it.
  t.is(blazen.TractOptions, undefined);
  t.is(blazen.JsTractOptions, undefined);
});

test("Tract typed bindings · TractOptions object literal is accepted by TractEmbedModel.create", (t) => {
  if (!HAS_TRACT) {
    t.pass("embed-tract feature not built");
    return;
  }

  // JsTractOptions is a structural interface: every field is optional.
  // create() is synchronous; pass a fully-populated literal to verify the
  // type surface accepts every documented field. We attach a no-op handler
  // to avoid an unhandled-rejection warning if the model loader fails.
  let model;
  try {
    model = blazen.TractEmbedModel.create({
      modelName: "BGESmallENV15",
      cacheDir: "/tmp/blazen-tract-cache",
      maxBatchSize: 32,
      showDownloadProgress: false,
    });
  } catch {
    // Model load may fail in offline / restricted environments. The shape
    // check is the important part; the runtime call is exercised by the
    // env-gated live test below.
    t.pass("model load failed in offline/restricted environment");
    return;
  }
  t.truthy(model, "expected a TractEmbedModel instance");
  t.is(typeof model.modelId, "string");
  t.is(typeof model.dimensions, "number");
});

test("Tract typed bindings · TractEmbedModel class shape (static create + instance methods/getters)", (t) => {
  if (!HAS_TRACT) {
    t.pass("embed-tract feature not built");
    return;
  }
  t.is(typeof blazen.TractEmbedModel, "function");
  t.is(typeof blazen.TractEmbedModel.create, "function");

  const proto = blazen.TractEmbedModel.prototype;

  // `embed` is an async method.
  t.is(
    typeof proto.embed,
    "function",
    "TractEmbedModel.prototype.embed must be a function",
  );

  // `modelId` and `dimensions` are getters, not methods.
  for (const name of ["modelId", "dimensions"]) {
    const desc = Object.getOwnPropertyDescriptor(proto, name);
    t.truthy(desc, `${name} getter must exist on prototype`);
    t.is(typeof desc.get, "function");
  }
});

test("Tract typed bindings · TractResponse is an object-literal interface (no runtime class)", (t) => {
  if (!HAS_TRACT) {
    t.pass("embed-tract feature not built");
    return;
  }
  // `JsTractResponse` is a TypeScript interface; the value yielded by
  // `embed()` is a plain object with `embeddings` and `model` fields.
  // Confirm there is no accidental class export shadowing it.
  t.is(blazen.TractResponse, undefined);
  t.is(blazen.JsTractResponse, undefined);
});

test("Tract typed bindings · TractError is an exception subclass of ProviderError", (t) => {
  if (!HAS_TRACT) {
    t.pass("embed-tract feature not built");
    return;
  }
  // The typed error tree is declared in `index.d.ts` but is only present in
  // `index.js` once the runtime error-wiring lands. Skip gracefully when
  // the class is absent so this smoke test can run on partial builds.
  if (typeof blazen.TractError !== "function") {
    t.pass("TractError runtime class not present");
    return;
  }

  // Must be a constructor whose instances are Errors.
  const err = new blazen.TractError("boom");
  t.truthy(err instanceof Error, "TractError instances must be Errors");
  t.truthy(
    err instanceof blazen.TractError,
    "TractError instances must be TractError",
  );

  // And TractError itself extends ProviderError when that base is
  // present (only true for the standard typed-error tree).
  if (typeof blazen.ProviderError === "function") {
    t.truthy(
      blazen.TractError.prototype instanceof blazen.ProviderError,
      "TractError must extend ProviderError",
    );
    t.truthy(
      err instanceof blazen.ProviderError,
      "TractError instance must also be a ProviderError",
    );
  }
});

// Live-model checks. Per the project's sync-only Node-test rule we cannot
// actually await `model.embed(...)` here, so we restrict ourselves to verifying
// that the env-gated path constructs the model and exposes the expected shape.
// Real end-to-end embedding is exercised by the Python smoke tests.
const TLive = TRACT_ENABLED ? test : test.skip;

TLive("Tract embed live (env-gated) · TractEmbedModel.create returns a model with positive dimensions", (t) => {
  if (!HAS_TRACT) {
    t.pass("embed-tract feature not built");
    return;
  }

  const model = blazen.TractEmbedModel.create();
  t.truthy(model, "create() must return a TractEmbedModel instance");
  t.truthy(
    typeof model.modelId === "string" && model.modelId.length > 0,
    "expected a non-empty model ID",
  );
  t.truthy(
    typeof model.dimensions === "number" && model.dimensions > 0,
    "expected positive dimensions",
  );
});

TLive("Tract embed live (env-gated) · TractEmbedModel.embed returns a thenable when invoked", (t) => {
  if (!HAS_TRACT) {
    t.pass("embed-tract feature not built");
    return;
  }

  const model = blazen.TractEmbedModel.create();
  // We cannot await in a sync test body. Just confirm `embed` returns a
  // Promise-shaped value; we attach a best-effort no-op handler so an
  // eventual rejection does not surface as an unhandled-rejection warning.
  const result = model.embed(["hello world"]);
  t.truthy(result, "embed() must return a value");
  t.is(typeof result.then, "function", "embed() must return a Promise");
  if (typeof result.catch === "function") {
    result.catch(() => {});
  }
});
